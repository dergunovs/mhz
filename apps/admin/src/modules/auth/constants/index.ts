export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
