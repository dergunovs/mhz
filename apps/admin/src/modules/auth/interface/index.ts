export interface ILoginFormData {
  email: string;
  password: string;
}
