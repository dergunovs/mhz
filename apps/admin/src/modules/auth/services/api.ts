import { api, setAuth } from 'mhz-helpers';

import {
  API_AUTH_CHECK,
  API_AUTH_LOGIN,
  API_AUTH_SETUP,
  ILoginData,
  ISignUpData,
  IBaseReply,
  IUserToken,
} from 'mhz-contracts';

export async function checkAuthApi() {
  const { data } = await api.get<IBaseReply>(API_AUTH_CHECK);

  if (data.message) setAuth(true);

  return data;
}
export async function setupApi(formData: ISignUpData) {
  const { data } = await api.post<IBaseReply>(API_AUTH_SETUP, formData);

  return data;
}

export async function loginApi(formData: ILoginData) {
  const { data } = await api.post<IUserToken>(API_AUTH_LOGIN, formData);

  return data;
}
