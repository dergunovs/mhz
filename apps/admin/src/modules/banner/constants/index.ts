export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
