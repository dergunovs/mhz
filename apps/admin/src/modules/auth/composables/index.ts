import { readonly, ref } from 'vue';
import { useRouter } from 'vue-router';

import { setAuthHeader, deleteAuthHeader } from '@/common/services/api';
import { URL_LOGIN } from '@/auth/constants';
import { URL_MAIN } from '@/common/constants';

const isAuthLocal = ref(false);

export const isAuth = readonly(isAuthLocal);

export function setAuth(auth: boolean) {
  isAuthLocal.value = auth;
}

export function logout() {
  deleteCookieToken();
  deleteAuthHeader();
  window.location.href = `${URL_LOGIN}?logout=1`;
}

export function setCookieToken(token: string) {
  document.cookie = `mhz_token=${token};Secure;samesite=strict;`;
}

export function getCookieToken(): string | undefined {
  const { mhz_token } = Object.fromEntries(document.cookie.split('; ').map((v) => v.split('=')));

  return mhz_token;
}

export function deleteCookieToken() {
  document.cookie = `mhz_token=;expires=${new Date(0).toUTCString()}`;
}

export function useAuth() {
  const router = useRouter();

  function login(token: string) {
    setCookieToken(token);
    setAuthHeader(token);
    setAuth(true);

    router.push(URL_MAIN);
  }

  return {
    login,
  };
}
