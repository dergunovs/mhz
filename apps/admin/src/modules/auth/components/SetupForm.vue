<template>
  <div>
    <ImageLogo />

    <h2>Add first manager</h2>

    <form @submit.prevent="submit" :class="$style.form">
      <UiField label="First name" isRequired :error="error('firstName')">
        <UiInput v-model="formData.firstName" isFocus />
      </UiField>

      <UiField label="Last name" isRequired :error="error('lastName')">
        <UiInput v-model="formData.lastName" />
      </UiField>

      <UiField label="E-mail" isRequired :error="error('email')">
        <UiInput v-model="formData.email" />
      </UiField>

      <UiField label="Password" isRequired :error="error('password')">
        <UiInput v-model="formData.password" type="password" />
      </UiField>

      <UiButton type="submit">Submit</UiButton>
    </form>
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import { useRouter } from 'vue-router';

import { UiButton, UiField, UiInput, toast } from 'mhz-ui';
import { useValidator, required, email } from 'mhz-helpers';
import { ISignUpData } from 'mhz-contracts';

import ImageLogo from '@/layout/icons/logo.svg';
import { setup } from '@/auth/services';
import { URL_LOGIN } from '@/auth/constants';

const router = useRouter();

const formData = ref<ISignUpData>({
  firstName: '',
  lastName: '',
  email: '',
  password: '',
});

const { mutate: mutateSetup } = setup({
  onSuccess: () => {
    toast.success('Manager successfully added!');
    router.push(URL_LOGIN);
  },
});

const rules = computed(() => {
  return {
    firstName: required,
    lastName: required,
    email: [required, email],
    password: required,
  };
});

const { error, isValid } = useValidator(formData, rules);

function submit() {
  if (isValid()) mutateSetup(formData.value);
}
</script>

<style module lang="scss">
.form {
  display: flex;
  flex-direction: column;
  gap: 24px;
  width: 300px;
}
</style>
