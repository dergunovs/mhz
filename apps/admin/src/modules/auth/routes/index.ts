import { URL_LOGIN } from '@/auth/constants';

export const authRoutes = [{ path: URL_LOGIN, name: 'Login', component: () => import('@/auth/pages/LoginPage.vue') }];
