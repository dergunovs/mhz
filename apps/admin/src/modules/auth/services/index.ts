import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
