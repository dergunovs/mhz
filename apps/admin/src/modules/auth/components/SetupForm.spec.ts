import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { IBaseReply, ISignUpData } from 'mhz-contracts';
import { toast } from 'mhz-ui';
import { dataTest } from 'mhz-helpers';

import SetupForm from './SetupForm.vue';

import { mockMutationReply, wrapperFactory, router } from '@/common/test';
import * as authServices from '@/auth/services';
import { URL_LOGIN } from '@/auth/constants';

const FIRST_NAME = 'John';
const LAST_NAME = 'Doe';
const EMAIL = 'a@b.ru';
const PASSWORD = 'qwerty';

let onSuccessSetup: () => void;
const spyMutateSetup = vi.fn();

vi.spyOn(authServices, 'setup').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessSetup = options.onSuccess;

  return mockMutationReply<IBaseReply, ISignUpData>(spyMutateSetup);
});

const spyToastSuccess = vi.spyOn(toast, 'success');
const spyRouterPush = vi.spyOn(router, 'push');

const setupForm = dataTest('setup-form');
const setupFormFirstName = dataTest('setup-form-first-name');
const setupFormLastName = dataTest('setup-form-last-name');
const setupFormEmail = dataTest('setup-form-email');
const setupFormPassword = dataTest('setup-form-password');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(SetupForm, {});
});

enableAutoUnmount(afterEach);

describe('SetupForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(SetupForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('handles setup by form submit', async () => {
    expect(spyMutateSetup).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    await wrapper.findComponent(setupFormFirstName).setValue(FIRST_NAME);
    await wrapper.findComponent(setupFormLastName).setValue(LAST_NAME);
    await wrapper.findComponent(setupFormEmail).setValue(EMAIL);
    await wrapper.findComponent(setupFormPassword).setValue(PASSWORD);

    await wrapper.find(setupForm).trigger('submit');

    expect(spyMutateSetup).toBeCalledTimes(1);
    expect(spyMutateSetup).toBeCalledWith({
      firstName: FIRST_NAME,
      lastName: LAST_NAME,
      email: EMAIL,
      password: PASSWORD,
    });

    onSuccessSetup();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_LOGIN);
  });
});
