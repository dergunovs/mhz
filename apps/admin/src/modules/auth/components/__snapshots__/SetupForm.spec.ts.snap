// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SetupForm > matches snapshot 1`] = `
"<div>
  <anonymous-stub></anonymous-stub>
  <h2>Add first manager</h2>
  <form class="_form_87c99b" data-test="setup-form">
    <fieldset label="First name" isrequired=""><input type="text" modelvalue="" isfocus="" data-test="setup-form-first-name"></fieldset>
    <fieldset label="Last name" isrequired=""><input type="text" modelvalue="" data-test="setup-form-last-name"></fieldset>
    <fieldset label="E-mail" isrequired=""><input type="text" modelvalue="" data-test="setup-form-email"></fieldset>
    <fieldset label="Password" isrequired=""><input type="password" modelvalue="" data-test="setup-form-password"></fieldset><button type="submit">Submit</button>
  </form>
</div>"
`;
